midend/simplifyKey.cpp

```diff
@@ -58,12 +58,12 @@ const IR::Node* DoSimplifyKey::postorder(IR::KeyElement* element) {
     auto type = typeMap->getType(element->expression, true);
     auto decl = new IR::Declaration_Variable(tmp, type, nullptr);
     insertions->declarations.push_back(decl);
-    auto left = new IR::PathExpression(tmp);
+    auto left = new IR::PathExpression(type, new IR::Path(tmp));
```

Comment:
I am not sure how this helps.
In general, the type field of an expression is not maintained by the compiler. 
There is one single pass which is used to populate it, called ApplyTypesToExpressions, which is uses sometimes to avoid looking up types in the typeMap. But any subsequent changes to the IR could make the type in the Expression incorrect.
This change is correct, but won't guarantee that the type field is correct after any subsequent pass that modifies the code.